renderdoc/driver/shaders/dxbc/dxbc_debug.h

```diff
@@ -147,6 +147,31 @@ struct GlobalState
   std::map<SampleEvalCacheKey, ShaderVariable> sampleEvalCache;
 };
 
+class DebugAPIWrapper
+{
+public:
+  virtual void SetCurrentInstruction(uint32_t instruction) = 0;
+  virtual void AddDebugMessage(MessageCategory c, MessageSeverity sv, MessageSource src,
+                               std::string d) = 0;
+
+  virtual bool CalculateRcp(const ShaderVariable &input, ShaderVariable &output) = 0;
+  virtual bool CalculateRsq(const ShaderVariable &input, ShaderVariable &output) = 0;
+  virtual bool CalculateExp(const ShaderVariable &input, ShaderVariable &output) = 0;
+  virtual bool CalculateLog(const ShaderVariable &input, ShaderVariable &output) = 0;
+  virtual bool CalculateSinCos(const ShaderVariable &input, ShaderVariable &output1,
+                               ShaderVariable &output2) = 0;
+
+  virtual ShaderVariable GetSampleInfo(const DXBC::ASMOperation &op) = 0;
+  virtual ShaderVariable GetBufferInfo(const DXBC::ASMOperation &op) = 0;
+  virtual ShaderVariable GetResourceInfo(const DXBC::ASMOperation &op, uint32_t mipLevel,
+                                         int &dim) = 0;
+
+  virtual bool CalculateSampleGather(const DXBC::ASMOperation &op,
+                                     const std::vector<ShaderVariable> &srcOpers,
+                                     ShaderVariable ddxCalc, ShaderVariable ddyCalc,
+                                     ShaderVariable &output) = 0;
```

Comment:
This will be harder to split apart than the above `Get*Info` functions but I think it's worth doing. The debug engine should be responsible for decoding the opcode, extracting out the operands, and passing along the information about the type & binding of the resource and the texture operation needed. Then the implementation of this class should only be responsible for mapping that request to the actual API and generating/running/fetching the shader. At the moment these two classes are pretty coupled because a lot of the implementation of this function needs to then poke back into DXBC internals.

Ideally then the implementation of this class will have very little dependency on the DXBC namespace (aside from using some enums

Reply:
Hmm, I see what you mean, and I agree with the value of splitting this into several, more targeted functions for the DXBC code to call into. This will take a while to sort through. :)

Reply:
Actually similarly to above in reverse, I'm not meaning to split the function apart. I think it's fine to be merged because there's a fair bit of commonality between the paths. Instead I'm referring to the same thing again - the implementation of CalculateSampleGather still has to do a lot of dxbc-specific work to decode the right operands, iterate over declarations to find resource bindings, etc.

Imagine if in future D3D11 was updated to accept SPIR-V as a shader encoding (hallelujah!), so we have a spirv_debug.cpp which wants to call the same interface but with a completely different bytecode. We want this interface to be mostly agnostic to the actual shader encoding it's debugging, it should simply be a provider for functions that we can't emulate on the CPU. We don't have to take it to extremes right now - it's fine to still use some DXBC enums like the opcode that could easily be genericised later - but I'd rather have the bulk of that dependency broken here.

Reply:
I see what you mean now, thanks for clarifying.

Taking another look, a lot of CalculateSampleGather is setting up all of the strings for the program (texture & sampler declaration, return type, texcoords, etc...it's a big list). That work feels best kept in DXBC since it'd be most likely just duplicated to D3D12. This would reduce code duplication but would result in a function signature full of std::strings which is extremely "go ahead and try not to shoot off your foot" feeling.

I'm inclined to make the function signature a soup of data from the ASMOperation that is needed, but add helper functions in DXBC to create each of those strings. Does that sound like the best strategy to you?

Reply:
Having a 'derived' struct of just the configuration needed from ASMOperation is what I had in mind yes. I think it's fine to have the HLSL string generation in D3D11, even if it gets duplicated in D3D12. I'm not so much worried about a bit of duplication as I am with having this code separated but still entangled. I'd rather have all the D3D/HLSL stuff in the D3D11 project and all the DXBC stuff in the DXBC project first. Then if later we decide some code in the D3D11 codebase needs to be extracted out to share with D3D12 it may make sense to put that somewhere else, not in the DXBC project. There's already `renderdoc/driver/dx` - it's currently only used for the shared official headers but it could be used for storing common helpers.

I haven't gone through exactly what data would need to be passed, I think it'd be useful to codify that as a first step and see how much is actually left. It may boil down to not that much apart from the type/dimension of the resource and the type of the co-ordinates. If that leaves a couple of string lookup tables or helpers in D3D11 that's not a problem for now.

I'm not sure about this either because I have a look at it, but the D3D12 HLSL generation will have to be somewhat different anyway to handle the different binding model I think.

Reply:
Thanks. I didn't go as far as creating a struct for params for the SampleGather function, but despite the number of params, I think it's pretty sensible, and in doing this I found a bunch of code that could stay in DXBC.